app/vmalert/rule/alerting.go

```diff
@@ -304,10 +304,16 @@ func (ar *AlertingRule) execRange(ctx context.Context, start, end time.Time) ([]
 		return nil, err
 	}
 	var result []prompbmarshal.TimeSeries
+	holdAlertState := make(map[uint64]*notifier.Alert)
```

Comment:
I don't think this would work. The problem of losing the state is not within this function. The state is lost between consecutive calls to `execRange` by replay logic. 
For example, we have a time range from 00:00 to 23:00 with step `1h` and eval_interval `1m`. That means that `execRange` will be called 23 times, for each hour. And within each call, function will be processing up to 60 data samples per series. 
The fix should retain state between `execRange` calls, so alerting rule which has `for:3h` could have been triggered within 3 `execRange` calls.
I think, we should do some copy&pasting from `exec` function and utilize `rule.alerts` field as state storage.

Reply:
I think this should work, `holdAlertState` will be [stored](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/keep-alert-state-when-replay/app/vmalert/rule/alerting.go#L357) in alertingRule's active alerts at the end, please check test cases [here](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/keep-alert-state-when-replay/app/vmalert/rule/alerting_test.go#L407). And those alert states will only be stored and restored in next `execRange` call if the condition is [right](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/keep-alert-state-when-replay/app/vmalert/rule/alerting.go#L333).

And a real test case result here.
![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/39937150/5f868da7-d3d8-4c0f-b274-0cc688fb92b4)

Reply:
> Yeah, AlertingRule.alerts is already used to persist state here. So holdAlertState will be [stored](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/keep-alert-state-when-replay/app/vmalert/rule/alerting.go#L357) in AlertingRule.alerts at the end

Thanks! I didn't see it in the end. I'm sorry. But why can't we use `a.alerts` directly? Am I missing something subtle?

Reply:
Although, we should update description of this function as it mentiones Rules state remains unchanged

Reply:
>But why can't we use a.alerts directly? Am I missing something subtle?

Nothing subtle, but I think using a.alerts all the time is not convenient(it need to be clean up after all the series done with start timestamp) and understandable compare to having a noticeable `holdAlertState`.

>Although, we should update description of this function as it mentiones Rules state remains unchanged

Yes, haven't notice that, fixed.